src/wallet/wallet2.cpp

```diff
@@ -7763,16 +7780,18 @@ bool wallet2::tx_add_fake_output(std::vector<std::vector<tools::wallet2::get_out
   if (std::find(outs.back().begin(), outs.back().end(), item) != outs.back().end()) // don't add duplicates
     return false;
   // check the keys are valid
-  if (!rct::isInMainSubgroup(rct::pk2rct(output_public_key)))
+  if (valid_public_keys_cache.find(output_public_key) == valid_public_keys_cache.end() && !rct::isInMainSubgroup(rct::pk2rct(output_public_key)))
```

Comment:
> All commits LGTM except https://github.com/monero-project/monero/commit/bd740838ac64e8739c5a81e2c2e0a0773f2fe3a7

It does say `5.9 second -> 5.2 seconds`, so there is an improvement.

Reply:
I meant if that commit's improvement is because the sample test case is selecting duplicate fake outs in a single tx, then I would think the commit probably shouldn't be included because I think duplicate outs should actually be prevented instead..

Unless there is some other way that this would yield an improvement I'm not seeing

Reply:
IIRC this is because this function ends up called several times when selecting rings. I'll try it to make sure.

Reply:
Doesn't seem to. Possibly I made several txes in a row to average, I don't remember. In that case, the timing would be unrepresentative, unless one cancels a tx, or most of the outputs the wallet requests are unusable, in which case the wallets asks for more (I think this includes the previous set). Probably down to averaging.

Reply:
Actually, I quite possibly tested on a short chain on which I mined a few hundred blocks first, so that'd mean fake out sets with overlapping outputs as you suggested. Doesn't seem enough to cause 5.9 -> 5.2 though, but who knows.

Reply:
I didn't consider reusing across attempts could come into play here, fair enough. Your call on it. Approving as is